File: hushh-webapp/ios/App/App/Plugins/HushhVault/HushhVaultPlugin.swift
//
//  HushhVaultPlugin.swift
//  App
//
//  Hushh Vault - Encrypted Storage Plugin
//  Port of lib/vault/encrypt.ts with AES-256-GCM encryption
//
//  Uses CryptoKit for:
//  - PBKDF2 key derivation (100,000 iterations, SHA-256)
//  - AES-256-GCM encryption (12-byte nonce, 128-bit tag)
//
//  Cloud DB Access:
//  - Calls consent-protocol Cloud Run backend for vault operations
//

import Foundation
import Capacitor
import CryptoKit

// MARK: - Constants

private let PBKDF2_ITERATIONS = 100_000
private let KEY_SIZE_BYTES = 32      // 256 bits
private let SALT_SIZE_BYTES = 16     // 128 bits
private let NONCE_SIZE_BYTES = 12    // 96 bits (as per NIST recommendation)

// Cloud Run Backend URL for vault operations
private let CLOUD_BACKEND_URL = "https://consent-protocol-1006304528804.us-central1.run.app"

// MARK: - Plugin

@objc(HushhVaultPlugin)
public class HushhVaultPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "HushhVaultPlugin"
    public let jsName = "HushhVault"
    public let pluginMethods: [CAPPluginMethod] = [
        // Crypto methods
        CAPPluginMethod(name: "deriveKey", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "encryptData", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "decryptData", returnType: CAPPluginReturnPromise),
        // Cloud DB methods (uses VaultStorageManager for cloud/local abstraction)
        CAPPluginMethod(name: "hasVault", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getVault", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setupVault", returnType: CAPPluginReturnPromise),
        // Local storage stubs
        CAPPluginMethod(name: "storePreference", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPreferences", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "deletePreferences", returnType: CAPPluginReturnPromise),
    ]
    
    // MARK: - Cloud DB: Has Vault Check
    
    /// Check if user has a vault
    /// Uses VaultStorageManager which routes to cloud or local storage
    @objc func hasVault(_ call: CAPPluginCall) {
        guard let userId = call.getString("userId") else {
            call.reject("Missing userId")
            return
        }
        
        // Update auth token if provided
        if let authToken = call.getString("authToken") {
            VaultStorageManager.shared.setAuthToken(authToken)
        }
        
        Task {
            do {
                let exists = try await VaultStorageManager.shared.getStorage().hasVault(userId: userId)
                call.resolve(["exists": exists])
            } catch {
                call.reject("Failed to check vault: \(error.localizedDescription)")
            }
        }
    }
    
    // MARK: - Cloud DB: Get Vault Data
    
    /// Get encrypted vault key
    /// Uses VaultStorageManager for cloud/local abstraction
    @objc func getVault(_ call: CAPPluginCall) {
        guard let userId = call.getString("userId") else {
            call.reject("Missing userId")
            return
        }
        
        if let authToken = call.getString("authToken") {
            VaultStorageManager.shared.setAuthToken(authToken)
        }
        
        Task {
            do {
                let data = try await VaultStorageManager.shared.getStorage().getVault(userId: userId)
                call.resolve([
                    "authMethod": data.authMethod,
                    "encryptedVaultKey": data.encryptedVaultKey,
                    "salt": data.salt,
                    "iv": data.iv,
                    "recoveryEncryptedVaultKey": data.recoveryEncryptedVaultKey,
                    "recoverySalt": data.recoverySalt,
                    "recoveryIv": data.recoveryIv
                ])
            } catch CloudDBError.notFound {
                call.reject("Vault not found", "NOT_FOUND")
            } catch {
                call.reject("Failed to get vault: \(error.localizedDescription)")
            }
        }
    }
    
    // MARK: - Cloud DB: Setup Vault
    
    /// Store encrypted vault key
    /// Uses VaultStorageManager for cloud/local abstraction
    @objc func setupVault(_ call: CAPPluginCall) {
        guard let userId = call.getString("userId"),
              let encryptedVaultKey = call.getString("encryptedVaultKey"),
              let salt = call.getString("salt"),
              let iv = call.getString("iv"),
              let recoveryEncryptedVaultKey = call.getString("recoveryEncryptedVaultKey"),
              let recoverySalt = call.getString("recoverySalt"),
              let recoveryIv = call.getString("recoveryIv") else {
            call.reject("Missing required parameters")
            return
        }
        
        if let authToken = call.getString("authToken") {
            VaultStorageManager.shared.setAuthToken(authToken)
        }
        
        let data = VaultKeyInfo(
            authMethod: call.getString("authMethod") ?? "passphrase",
            encryptedVaultKey: encryptedVaultKey,
            salt: salt,
            iv: iv,
            recoveryEncryptedVaultKey: recoveryEncryptedVaultKey,
            recoverySalt: recoverySalt,
            recoveryIv: recoveryIv
        )
        
        Task {
            do {
                try await VaultStorageManager.shared.getStorage().setupVault(userId: userId, data: data)
                call.resolve(["success": true])
            } catch {
                call.reject("Failed to setup vault: \(error.localizedDescription)")
            }
        }
    }
    
    // MARK: - Derive Key
    
    /// Derive an encryption key from passphrase using PBKDF2
    /// Mirrors: PBKDF2 key derivation in encrypt.ts
    ///
    /// Parameters:
    /// - iterations: 100,000 (matches config.py)
    /// - hash: SHA-256
    /// - keyLength: 256 bits (32 bytes, output as 64-char hex)
    @objc func deriveKey(_ call: CAPPluginCall) {
        guard let passphrase = call.getString("passphrase") else {
            call.reject("Missing required parameter: passphrase")
            return
        }
        
        let iterations = call.getInt("iterations") ?? PBKDF2_ITERATIONS
        
        // Generate or use provided salt
        let saltBytes: Data
        if let saltHex = call.getString("salt") {
            saltBytes = Data(hexString: saltHex) ?? Data()
            if saltBytes.isEmpty {
                call.reject("Invalid salt hex string")
                return
            }
        } else {
            var randomBytes = [UInt8](repeating: 0, count: SALT_SIZE_BYTES)
            _ = SecRandomCopyBytes(kSecRandomDefault, SALT_SIZE_BYTES, &randomBytes)
            saltBytes = Data(randomBytes)
        }
        
        // Derive key using PBKDF2 with CommonCrypto
        guard let derivedKey = pbkdf2Derive(
            password: passphrase,
            salt: saltBytes,
            iterations: iterations,
            keyLength: KEY_SIZE_BYTES
        ) else {
            call.reject("Key derivation failed")
            return
        }
        
        // Convert to hex strings
        let keyHex = derivedKey.map { String(format: "%02x", $0) }.joined()
        let saltHex = saltBytes.map { String(format: "%02x", $0) }.joined()
        
        call.resolve([
            "keyHex": keyHex,
            "salt": saltHex
        ])
    }
    
    // MARK: - Encrypt Data
    
    /// Encrypt data using AES-256-GCM
    /// Mirrors: encryptData() in encrypt.ts
    ///
    /// Uses: 12-byte IV, returns ciphertext + IV + tag in base64
    @objc func encryptData(_ call: CAPPluginCall) {
        guard let plaintext = call.getString("plaintext"),
              let keyHex = call.getString("keyHex") else {
            call.reject("Missing required parameters: plaintext, keyHex")
            return
        }
        
        // Parse key from hex
        guard let keyData = Data(hexString: keyHex),
              keyData.count == KEY_SIZE_BYTES else {
            call.reject("Invalid key: must be 64-character hex string (256-bit)")
            return
        }
        
        let key = SymmetricKey(data: keyData)
        
        // Generate random nonce (12 bytes)
        var nonceBytes = [UInt8](repeating: 0, count: NONCE_SIZE_BYTES)
        _ = SecRandomCopyBytes(kSecRandomDefault, NONCE_SIZE_BYTES, &nonceBytes)
        
        guard let nonce = try? AES.GCM.Nonce(data: Data(nonceBytes)) else {
            call.reject("Failed to generate nonce")
            return
        }
        
        // Encrypt
        guard let plaintextData = plaintext.data(using: .utf8),
              let sealedBox = try? AES.GCM.seal(plaintextData, using: key, nonce: nonce) else {
            call.reject("Encryption failed")
            return
        }
        
        // Extract components
        // CryptoKit's sealedBox.ciphertext includes only ciphertext (no tag)
        // sealedBox.tag is the authentication tag
        let ciphertext = sealedBox.ciphertext
        let tag = sealedBox.tag
        let iv = Data(nonceBytes)
        
        call.resolve([
            "ciphertext": ciphertext.base64EncodedString(),
            "iv": iv.base64EncodedString(),
            "tag": tag.base64EncodedString(),
            "encoding": "base64",
            "algorithm": "aes-256-gcm"
        ])
    }
    
    // MARK: - Decrypt Data
    
    /// Decrypt data using AES-256-GCM
    /// Mirrors: decryptData() in encrypt.ts
    @objc func decryptData(_ call: CAPPluginCall) {
        guard let payload = call.getObject("payload"),
              let keyHex = call.getString("keyHex") else {
            call.reject("Missing required parameters: payload, keyHex")
            return
        }
        
        guard let ciphertextB64 = payload["ciphertext"] as? String,
              let ivB64 = payload["iv"] as? String,
              let tagB64 = payload["tag"] as? String else {
            call.reject("Invalid payload: missing ciphertext, iv, or tag")
            return
        }
        
        // Parse components from base64
        guard let ciphertext = Data(base64Encoded: ciphertextB64),
              let iv = Data(base64Encoded: ivB64),
              let tag = Data(base64Encoded: tagB64) else {
            call.reject("Invalid base64 encoding in payload")
            return
        }
        
        // Parse key from hex
        guard let keyData = Data(hexString: keyHex),
              keyData.count == KEY_SIZE_BYTES else {
            call.reject("Invalid key")
            return
        }
        
        let key = SymmetricKey(data: keyData)
        
        // Create nonce from IV
        guard let nonce = try? AES.GCM.Nonce(data: iv) else {
            call.reject("Invalid IV")
            return
        }
        
        // Reconstruct sealed box
        guard let sealedBox = try? AES.GCM.SealedBox(nonce: nonce, ciphertext: ciphertext, tag: tag) else {
            call.reject("Failed to reconstruct sealed box")
            return
        }
        
        // Decrypt
        guard let decryptedData = try? AES.GCM.open(sealedBox, using: key),
              let plaintext = String(data: decryptedData, encoding: .utf8) else {
            call.reject("Decryption failed - invalid key or corrupted data")
            return
        }
        
        call.resolve([
            "plaintext": plaintext
        ])
    }
    
    // MARK: - Store Preference (Stub for Phase 2 SQLCipher)
    
    /// Store encrypted preference in local database
    /// NOTE: Full SQLCipher implementation is Phase 2
    @objc func storePreference(_ call: CAPPluginCall) {
        // For now, store in UserDefaults (encrypted data only)
        // Phase 2 will implement SQLCipher
        guard let userId = call.getString("userId"),
              let domain = call.getString("domain"),
              let fieldName = call.getString("fieldName"),
              let data = call.getObject("data") else {
            call.reject("Missing required parameters")
            return
        }
        
        let key = "vault_\(userId)_\(domain)_\(fieldName)"
        
        // Serialize the encrypted payload
        if let jsonData = try? JSONSerialization.data(withJSONObject: data, options: []) {
            UserDefaults.standard.set(jsonData, forKey: key)
            call.resolve()
        } else {
            call.reject("Failed to serialize data")
        }
    }
    
    // MARK: - Get Preferences (Stub for Phase 2 SQLCipher)
    
    /// Retrieve preferences from local database
    /// NOTE: Full SQLCipher implementation is Phase 2
    @objc func getPreferences(_ call: CAPPluginCall) {
        guard let userId = call.getString("userId"),
              let domain = call.getString("domain") else {
            call.reject("Missing required parameters")
            return
        }
        
        // Search UserDefaults for matching keys
        let prefix = "vault_\(userId)_\(domain)_"
        var preferences: [[String: Any]] = []
        
        for key in UserDefaults.standard.dictionaryRepresentation().keys {
            if key.hasPrefix(prefix) {
                if let jsonData = UserDefaults.standard.data(forKey: key),
                   let data = try? JSONSerialization.jsonObject(with: jsonData) as? [String: Any] {
                    let fieldName = String(key.dropFirst(prefix.count))
                    preferences.append([
                        "userId": userId,
                        "domain": domain,
                        "fieldName": fieldName,
                        "data": data,
                        "createdAt": Date().timeIntervalSince1970 * 1000
                    ])
                }
            }
        }
        
        call.resolve([
            "preferences": preferences
        ])
    }
    
    // MARK: - Delete Preferences (Stub for Phase 2 SQLCipher)
    
    /// Delete preferences for a domain
    @objc func deletePreferences(_ call: CAPPluginCall) {
        guard let userId = call.getString("userId"),
              let domain = call.getString("domain") else {
            call.reject("Missing required parameters")
            return
        }
        
        let prefix = "vault_\(userId)_\(domain)_"
        
        for key in UserDefaults.standard.dictionaryRepresentation().keys {
            if key.hasPrefix(prefix) {
                UserDefaults.standard.removeObject(forKey: key)
            }
        }
        
        call.resolve()
    }
    
    // MARK: - Private Helpers
    
    /// PBKDF2 key derivation using CommonCrypto
    private func pbkdf2Derive(password: String, salt: Data, iterations: Int, keyLength: Int) -> Data? {
        guard let passwordData = password.data(using: .utf8) else {
            return nil
        }
        
        var derivedKey = [UInt8](repeating: 0, count: keyLength)
        
        let status = passwordData.withUnsafeBytes { passwordBytes in
            salt.withUnsafeBytes { saltBytes in
                CCKeyDerivationPBKDF(
                    CCPBKDFAlgorithm(kCCPBKDF2),
                    passwordBytes.baseAddress?.assumingMemoryBound(to: Int8.self),
                    passwordData.count,
                    saltBytes.baseAddress?.assumingMemoryBound(to: UInt8.self),
                    salt.count,
                    CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA256),
                    UInt32(iterations),
                    &derivedKey,
                    keyLength
                )
            }
        }
        
        guard status == kCCSuccess else {
            return nil
        }
        
        return Data(derivedKey)
    }
}

// MARK: - CommonCrypto Import

import CommonCrypto

// MARK: - Data Hex Extension

private extension Data {
    init?(hexString: String) {
        let hex = hexString.replacingOccurrences(of: " ", with: "")
        guard hex.count % 2 == 0 else { return nil }
        
        var data = Data(capacity: hex.count / 2)
        var index = hex.startIndex
        
        while index < hex.endIndex {
            let nextIndex = hex.index(index, offsetBy: 2)
            guard let byte = UInt8(hex[index..<nextIndex], radix: 16) else {
                return nil
            }
            data.append(byte)
            index = nextIndex
        }
        
        self = data
    }
}

File: hushh-webapp/android/app/src/main/java/com/hushh/pda/plugins/HushhVault/HushhVaultPlugin.kt
package com.hushh.pda.plugins.HushhVault

import android.util.Base64
import android.util.Log
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec

/**
 * Hushh Vault Plugin - Encryption + Cloud DB Proxy
 * Port of lib/vault/encrypt.ts and iOS HushhVaultPlugin.swift
 *
 * Uses: AES-256-GCM, PBKDF2 with 100,000 iterations
 */
@CapacitorPlugin(name = "HushhVault")
class HushhVaultPlugin : Plugin() {

    private val TAG = "HushhVault"
    private val httpClient = OkHttpClient()

    // Default Cloud Run backend URL (fallback if not provided by JS layer)
    private val defaultBackendUrl = "https://consent-protocol-1006304528804.us-central1.run.app"

    // ==================== Derive Key ====================

    @PluginMethod
    fun deriveKey(call: PluginCall) {
        val passphrase = call.getString("passphrase")
        if (passphrase == null) {
            call.reject("Missing required parameter: passphrase")
            return
        }

        val saltString = call.getString("salt")
        val iterations = call.getInt("iterations") ?: 100000

        try {
            // Generate or use provided salt
            val salt = if (saltString != null) {
                hexStringToByteArray(saltString)
            } else {
                ByteArray(32).also { SecureRandom().nextBytes(it) }
            }

            // PBKDF2 key derivation
            val factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
            val spec = PBEKeySpec(passphrase.toCharArray(), salt, iterations, 256)
            val key = factory.generateSecret(spec)

            val keyHex = key.encoded.toHexString()
            val saltHex = salt.toHexString()

            Log.d(TAG, "✅ [HushhVault] Key derived successfully")

            call.resolve(JSObject().apply {
                put("keyHex", keyHex)
                put("salt", saltHex)
            })
        } catch (e: Exception) {
            Log.e(TAG, "❌ [HushhVault] Key derivation failed: ${e.message}")
            call.reject("Key derivation failed: ${e.message}")
        }
    }

    // ==================== Encrypt Data ====================

    @PluginMethod
    fun encryptData(call: PluginCall) {
        val plaintext = call.getString("plaintext")
        val keyHex = call.getString("keyHex")

        if (plaintext == null || keyHex == null) {
            call.reject("Missing required parameters: plaintext, keyHex")
            return
        }

        try {
            val key = hexStringToByteArray(keyHex)
            val iv = ByteArray(12).also { SecureRandom().nextBytes(it) }

            val cipher = Cipher.getInstance("AES/GCM/NoPadding")
            val secretKey = SecretKeySpec(key, "AES")
            val gcmSpec = GCMParameterSpec(128, iv)
            cipher.init(Cipher.ENCRYPT_MODE, secretKey, gcmSpec)

            val encrypted = cipher.doFinal(plaintext.toByteArray(Charsets.UTF_8))

            // Split ciphertext and tag (last 16 bytes is the auth tag in GCM)
            val ciphertext = encrypted.dropLast(16).toByteArray()
            val tag = encrypted.takeLast(16).toByteArray()

            Log.d(TAG, "✅ [HushhVault] Data encrypted successfully")

            call.resolve(JSObject().apply {
                put("ciphertext", Base64.encodeToString(ciphertext, Base64.NO_WRAP))
                put("iv", Base64.encodeToString(iv, Base64.NO_WRAP))
                put("tag", Base64.encodeToString(tag, Base64.NO_WRAP))
                put("encoding", "base64")
                put("algorithm", "aes-256-gcm")
            })
        } catch (e: Exception) {
            Log.e(TAG, "❌ [HushhVault] Encryption failed: ${e.message}")
            call.reject("Encryption failed: ${e.message}")
        }
    }

    // ==================== Decrypt Data ====================

    @PluginMethod
    fun decryptData(call: PluginCall) {
        val payload = call.getObject("payload")
        val keyHex = call.getString("keyHex")

        if (payload == null || keyHex == null) {
            call.reject("Missing required parameters: payload, keyHex")
            return
        }

        try {
            val ciphertextStr = payload.getString("ciphertext")
            val ivStr = payload.getString("iv")
            val tagStr = payload.getString("tag")

            if (ciphertextStr == null || ivStr == null || tagStr == null) {
                call.reject("Invalid payload: missing ciphertext, iv, or tag")
                return
            }

            val key = hexStringToByteArray(keyHex)
            val ciphertext = Base64.decode(ciphertextStr, Base64.DEFAULT)
            val iv = Base64.decode(ivStr, Base64.DEFAULT)
            val tag = Base64.decode(tagStr, Base64.DEFAULT)

            // Combine ciphertext + tag for GCM
            val combined = ciphertext + tag

            val cipher = Cipher.getInstance("AES/GCM/NoPadding")
            val secretKey = SecretKeySpec(key, "AES")
            val gcmSpec = GCMParameterSpec(128, iv)
            cipher.init(Cipher.DECRYPT_MODE, secretKey, gcmSpec)

            val decrypted = cipher.doFinal(combined)
            val plaintext = String(decrypted, Charsets.UTF_8)

            Log.d(TAG, "✅ [HushhVault] Data decrypted successfully")

            call.resolve(JSObject().put("plaintext", plaintext))
        } catch (e: Exception) {
            Log.e(TAG, "❌ [HushhVault] Decryption failed: ${e.message}")
            call.reject("Decryption failed: ${e.message}")
        }
    }

    // ==================== Cloud DB Methods ====================

    @PluginMethod
    fun hasVault(call: PluginCall) {
        val userId = call.getString("userId")
        if (userId == null) {
            call.reject("Missing required parameter: userId")
            return
        }

        val authToken = call.getString("authToken")
        val backendUrl = call.getString("backendUrl") ?: defaultBackendUrl
        // iOS uses POST to /db/vault/check with JSON body
        val url = "$backendUrl/db/vault/check"
        
        Log.d(TAG, "🔐 [hasVault] Checking vault for userId: $userId")
        Log.d(TAG, "🔐 [hasVault] URL: $url")
        Log.d(TAG, "🔐 [hasVault] AuthToken present: ${authToken != null}")

        Thread {
            try {
                // Create JSON body like iOS does
                val jsonBody = JSONObject().apply {
                    put("userId", userId)
                }
                val requestBody = jsonBody.toString().toRequestBody("application/json".toMediaType())
                
                val requestBuilder = Request.Builder()
                    .url(url)
                    .post(requestBody)
                    .addHeader("Content-Type", "application/json")

                if (authToken != null) {
                    requestBuilder.addHeader("Authorization", "Bearer $authToken")
                }

                val response = httpClient.newCall(requestBuilder.build()).execute()
                val responseCode = response.code
                val body = response.body?.string() ?: "{}"
                
                Log.d(TAG, "🔐 [hasVault] Response code: $responseCode")
                Log.d(TAG, "🔐 [hasVault] Response body: $body")
                
                val json = JSONObject(body)
                
                // Backend returns hasVault
                val exists = json.optBoolean("hasVault", false)
                
                Log.d(TAG, "🔐 [hasVault] Parsed exists: $exists")

                activity.runOnUiThread {
                    call.resolve(JSObject().put("exists", exists))
                }
            } catch (e: Exception) {
                Log.e(TAG, "❌ [hasVault] Error: ${e.message}")
                activity.runOnUiThread {
                    call.reject("Failed to check vault: ${e.message}")
                }
            }
        }.start()
    }

    @PluginMethod
    fun getVault(call: PluginCall) {
        val userId = call.getString("userId")
        if (userId == null) {
            call.reject("Missing required parameter: userId")
            return
        }

        val authToken = call.getString("authToken")
        val backendUrl = call.getString("backendUrl") ?: defaultBackendUrl
        // iOS uses POST to /db/vault/get with JSON body
        val url = "$backendUrl/db/vault/get"

        Thread {
            try {
                // Create JSON body like iOS does
                val jsonBody = JSONObject().apply {
                    put("userId", userId)
                }
                val requestBody = jsonBody.toString().toRequestBody("application/json".toMediaType())
                
                val requestBuilder = Request.Builder()
                    .url(url)
                    .post(requestBody)
                    .addHeader("Content-Type", "application/json")

                if (authToken != null) {
                    requestBuilder.addHeader("Authorization", "Bearer $authToken")
                }

                val response = httpClient.newCall(requestBuilder.build()).execute()
                val body = response.body?.string() ?: "{}"
                val json = JSONObject(body)

                activity.runOnUiThread {
                    call.resolve(JSObject().apply {
                        put("authMethod", json.optString("authMethod", "passphrase"))
                        put("encryptedVaultKey", json.optString("encryptedVaultKey", ""))
                        put("salt", json.optString("salt", ""))
                        put("iv", json.optString("iv", ""))
                        put("recoveryEncryptedVaultKey", json.optString("recoveryEncryptedVaultKey", ""))
                        put("recoverySalt", json.optString("recoverySalt", ""))
                        put("recoveryIv", json.optString("recoveryIv", ""))
                    })
                }
            } catch (e: Exception) {
                activity.runOnUiThread {
                    call.reject("Failed to get vault: ${e.message}")
                }
            }
        }.start()
    }

    @PluginMethod
    fun setupVault(call: PluginCall) {
        val userId = call.getString("userId")
        val encryptedVaultKey = call.getString("encryptedVaultKey")
        val salt = call.getString("salt")
        val iv = call.getString("iv")
        val recoveryEncryptedVaultKey = call.getString("recoveryEncryptedVaultKey")
        val recoverySalt = call.getString("recoverySalt")
        val recoveryIv = call.getString("recoveryIv")

        if (userId == null || encryptedVaultKey == null || salt == null || iv == null) {
            call.reject("Missing required parameters")
            return
        }

        val authToken = call.getString("authToken")
        val backendUrl = call.getString("backendUrl") ?: defaultBackendUrl
        val authMethod = call.getString("authMethod") ?: "passphrase"

        Thread {
            try {
                val json = JSONObject().apply {
                    put("userId", userId)
                    put("authMethod", authMethod)
                    put("encryptedVaultKey", encryptedVaultKey)
                    put("salt", salt)
                    put("iv", iv)
                    put("recoveryEncryptedVaultKey", recoveryEncryptedVaultKey ?: "")
                    put("recoverySalt", recoverySalt ?: "")
                    put("recoveryIv", recoveryIv ?: "")
                }

                val requestBody = json.toString().toRequestBody("application/json".toMediaType())
                val requestBuilder = Request.Builder()
                    .url("$backendUrl/db/vault/setup")
                    .post(requestBody)
                    .addHeader("Content-Type", "application/json")

                if (authToken != null) {
                    requestBuilder.addHeader("Authorization", "Bearer $authToken")
                }

                val response = httpClient.newCall(requestBuilder.build()).execute()
                val success = response.isSuccessful

                activity.runOnUiThread {
                    call.resolve(JSObject().put("success", success))
                }
            } catch (e: Exception) {
                activity.runOnUiThread {
                    call.reject("Failed to setup vault: ${e.message}")
                }
            }
        }.start()
    }

    // ==================== Domain Data Methods ====================
    // These call Cloud Run backend to get/store encrypted user data

    @PluginMethod
    fun getFoodPreferences(call: PluginCall) {
        val userId = call.getString("userId")
        if (userId == null) {
            call.reject("Missing required parameter: userId")
            return
        }

        val authToken = call.getString("authToken")
        val sessionToken = call.getString("sessionToken")
        val backendUrl = call.getString("backendUrl") ?: defaultBackendUrl
        val url = "$backendUrl/db/food/get"

        Log.d(TAG, "🍽️ [getFoodPreferences] Fetching food data for userId: $userId")

        Thread {
            try {
                val jsonBody = JSONObject().apply {
                    put("userId", userId)
                }
                val requestBody = jsonBody.toString().toRequestBody("application/json".toMediaType())
                
                val requestBuilder = Request.Builder()
                    .url(url)
                    .post(requestBody)
                    .addHeader("Content-Type", "application/json")

                if (authToken != null) {
                    requestBuilder.addHeader("Authorization", "Bearer $authToken")
                }
                if (sessionToken != null) {
                    requestBuilder.addHeader("X-Session-Token", sessionToken)
                }

                val response = httpClient.newCall(requestBuilder.build()).execute()
                val responseCode = response.code
                val body = response.body?.string() ?: "{}"
                
                Log.d(TAG, "🍽️ [getFoodPreferences] Response code: $responseCode")

                if (responseCode == 404) {
                    // No data found - return empty preferences
                    activity.runOnUiThread {
                        call.resolve(JSObject().apply {
                            put("domain", "food")
                            put("preferences", JSONObject.NULL)
                        })
                    }
                    return@Thread
                }

                if (responseCode != 200) {
                    Log.e(TAG, "❌ [getFoodPreferences] Error: $body")
                    activity.runOnUiThread {
                        call.reject("Failed to get food preferences: HTTP $responseCode")
                    }
                    return@Thread
                }

                val json = JSONObject(body)
                val preferences = json.optJSONObject("preferences")

                activity.runOnUiThread {
                    call.resolve(JSObject().apply {
                        put("domain", "food")
                        put("preferences", preferences ?: JSONObject.NULL)
                    })
                }
            } catch (e: Exception) {
                Log.e(TAG, "❌ [getFoodPreferences] Error: ${e.message}")
                activity.runOnUiThread {
                    call.reject("Failed to get food preferences: ${e.message}")
                }
            }
        }.start()
    }

    @PluginMethod
    fun getProfessionalData(call: PluginCall) {
        val userId = call.getString("userId")
        if (userId == null) {
            call.reject("Missing required parameter: userId")
            return
        }

        val authToken = call.getString("authToken")
        val sessionToken = call.getString("sessionToken")
        val backendUrl = call.getString("backendUrl") ?: defaultBackendUrl
        val url = "$backendUrl/db/professional/get"

        Log.d(TAG, "💼 [getProfessionalData] Fetching professional data for userId: $userId")

        Thread {
            try {
                val jsonBody = JSONObject().apply {
                    put("userId", userId)
                }
                val requestBody = jsonBody.toString().toRequestBody("application/json".toMediaType())
                
                val requestBuilder = Request.Builder()
                    .url(url)
                    .post(requestBody)
                    .addHeader("Content-Type", "application/json")

                if (authToken != null) {
                    requestBuilder.addHeader("Authorization", "Bearer $authToken")
                }
                if (sessionToken != null) {
                    requestBuilder.addHeader("X-Session-Token", sessionToken)
                }

                val response = httpClient.newCall(requestBuilder.build()).execute()
                val responseCode = response.code
                val body = response.body?.string() ?: "{}"
                
                Log.d(TAG, "💼 [getProfessionalData] Response code: $responseCode")

                if (responseCode == 404) {
                    // No data found - return empty preferences
                    activity.runOnUiThread {
                        call.resolve(JSObject().apply {
                            put("domain", "professional")
                            put("preferences", JSONObject.NULL)
                        })
                    }
                    return@Thread
                }

                if (responseCode != 200) {
                    Log.e(TAG, "❌ [getProfessionalData] Error: $body")
                    activity.runOnUiThread {
                        call.reject("Failed to get professional data: HTTP $responseCode")
                    }
                    return@Thread
                }

                val json = JSONObject(body)
                val preferences = json.optJSONObject("preferences")

                activity.runOnUiThread {
                    call.resolve(JSObject().apply {
                        put("domain", "professional")
                        put("preferences", preferences ?: JSONObject.NULL)
                    })
                }
            } catch (e: Exception) {
                Log.e(TAG, "❌ [getProfessionalData] Error: ${e.message}")
                activity.runOnUiThread {
                    call.reject("Failed to get professional data: ${e.message}")
                }
            }
        }.start()
    }

    @PluginMethod
    fun storePreferencesToCloud(call: PluginCall) {
        val userId = call.getString("userId")
        val domain = call.getString("domain")
        val fieldName = call.getString("fieldName")
        val ciphertext = call.getString("ciphertext")
        val iv = call.getString("iv")
        val tag = call.getString("tag")

        if (userId == null || domain == null || fieldName == null || 
            ciphertext == null || iv == null || tag == null) {
            call.reject("Missing required parameters")
            return
        }

        val authToken = call.getString("authToken")
        val consentToken = call.getString("consentToken")
        val backendUrl = call.getString("backendUrl") ?: defaultBackendUrl
        val url = "$backendUrl/db/$domain/store"

        Log.d(TAG, "💾 [storePreferencesToCloud] Storing $fieldName for userId: $userId")

        Thread {
            try {
                val jsonBody = JSONObject().apply {
                    put("userId", userId)
                    put("fieldName", fieldName)
                    put("ciphertext", ciphertext)
                    put("iv", iv)
                    put("tag", tag)
                    if (consentToken != null) {
                        put("consentToken", consentToken)
                    }
                }
                val requestBody = jsonBody.toString().toRequestBody("application/json".toMediaType())
                
                val requestBuilder = Request.Builder()
                    .url(url)
                    .post(requestBody)
                    .addHeader("Content-Type", "application/json")

                if (authToken != null) {
                    requestBuilder.addHeader("Authorization", "Bearer $authToken")
                }

                val response = httpClient.newCall(requestBuilder.build()).execute()
                val responseCode = response.code
                val body = response.body?.string() ?: "{}"
                
                Log.d(TAG, "💾 [storePreferencesToCloud] Response code: $responseCode")

                if (responseCode != 200 && responseCode != 201) {
                    Log.e(TAG, "❌ [storePreferencesToCloud] Error: $body")
                    activity.runOnUiThread {
                        call.reject("Failed to store preferences: HTTP $responseCode")
                    }
                    return@Thread
                }

                activity.runOnUiThread {
                    call.resolve(JSObject().apply {
                        put("success", true)
                        put("field", fieldName)
                    })
                }
            } catch (e: Exception) {
                Log.e(TAG, "❌ [storePreferencesToCloud] Error: ${e.message}")
                activity.runOnUiThread {
                    call.reject("Failed to store preferences: ${e.message}")
                }
            }
        }.start()
    }

    // ==================== Preference Storage (Placeholder for SQLCipher) ====================

    @PluginMethod
    fun storePreference(call: PluginCall) {
        // Placeholder - will be implemented with SQLCipher
        call.resolve()
    }

    @PluginMethod
    fun getPreferences(call: PluginCall) {
        // Placeholder - will be implemented with SQLCipher
        call.resolve(JSObject().put("preferences", JSObject()))
    }

    @PluginMethod
    fun deletePreferences(call: PluginCall) {
        // Placeholder - will be implemented with SQLCipher
        call.resolve()
    }

    // ==================== Utility Functions ====================

    private fun ByteArray.toHexString(): String = joinToString("") { "%02x".format(it) }

    private fun hexStringToByteArray(hex: String): ByteArray {
        val result = ByteArray(hex.length / 2)
        for (i in result.indices) {
            result[i] = hex.substring(i * 2, i * 2 + 2).toInt(16).toByte()
        }
        return result
    }
}
